### PR TITLE
Feat/mfa

### PR DESCRIPTION
**Pull request by nguyentb**
Changes for commit:
	modified:   packages/itmat-commons/src/graphql/appUsers.ts
	modified:   packages/itmat-commons/src/graphql/user.ts
	modified:   packages/itmat-commons/src/models/user.ts
	modified:   packages/itmat-interface/src/graphql/core/userCore.ts
	modified:   packages/itmat-interface/src/graphql/resolvers/userResolvers.ts
	modified:   packages/itmat-interface/src/graphql/schema.ts

	modified:   packages/itmat-ui-react/src/components/login/login.tsx
	modified:   packages/itmat-ui-react/src/components/users/userDetails.tsx

	modified:   packages/itmat-utils/src/databaseSetup/collectionsAndIndexes.ts
	modified:   packages/itmat-utils/src/databaseSetup/seed/users.js
	modified:   packages/itmat-utils/src/models/user.ts

Untracked files:
	packages/itmat-interface/src/utils/mfa.js
	packages/itmat-utils/src/databaseSetup/collectionsAndIndexes.js
	packages/itmat-utils/src/databaseSetup/seed/users.bak.js
	yarn.lock

### CODE REVIEW

**Comment by fguitton** [comment]
@nguyentb This branch is out of sync with the base develop. Also as I know that the react-script issue has been fixed in branch develop as @sou-chon push some changes, you might want to rebase you single commit directly off of there. This will avoid PR pollution with my commits which are not strictly related to the MFA 😉

**Comment by sou-chon** [comment]
Hi could you make sure the tests are passing first?

**Comment by nguyentb** [comment]
hi Chon, can you revise the pull request as I modify the tests accordingly

**Comment by sou-chon** [comment]
> hi Chon, can you revise the pull request as I modify the tests accordingly

Hi, the tests are still not passing. Could you look into that?

**Comment by sou-chon** [comment]
Also I think there's a codependency issue with graphql. Some packages like ws-subscription requires graphql 14. Could you bump graphql version back to 14?

File: packages/itmat-commons/src/models/study.ts [hunk 1]
```diff
@@ -27,7 +29,7 @@ export interface IRole {
     studyId: string;
     name: string;
     permissions: string[];
-    users: string[];
+    users: Models.UserModels.IUser[];
```
Comment (sou-chon) [inline packages/itmat-commons/src/models/study.ts:32]: I think this would not be correct. Why did you change this? This does not correspond to what is stored in the mongodb. and you didn't map it back

File: packages/itmat-interface/src/graphql/resolvers/userResolvers.ts [hunk 2]
```diff
@@ -101,6 +102,12 @@ export const userResolvers = {
             delete result.password;
             delete result.deleted;
 
+			// validate the TOTP
+			const totpValidated = mfa.verifyTOTP(args.totp, result.otpSecret);
+			if (!totpValidated) {
+				throw new UserInputError('Incorrect TOTP. Obtain the TOTP using Google Authenticator app.');
+			}
+
```
Comment (sou-chon) [inline packages/itmat-interface/src/graphql/resolvers/userResolvers.ts:110]: could you re-indent this

File: packages/itmat-interface/src/utils/mfa.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+   }
+
+   for (let errorWindow = -window; errorWindow <= +window; errorWindow++) {
+	  const totp = generateTOTP(secret, errorWindow);
+	  if (parseInt(token) === totp) {
+	     return true;
+	  }
+   }
+   return false;
+}
```
Comment (sou-chon) [inline packages/itmat-interface/src/utils/mfa.js:63]: could you convert this to typescript

File: packages/itmat-commons/src/models/study.ts (hunk 1 above)
Comment (sou-chon) [reply packages/itmat-commons/src/models/study.ts:32]: @fguitton 
ok. I intended this IRole to be like in mongodb entry. This IUser[] field has an extra graphql resolver on Role type that resolves the strings into IUser. The type is defined in itmat-interface/graphql/schema.js. 

Are you using this interface in the UI / any code that requires users key to be IUser[]?  If needed, could you make something like

"""
interface IRoleFromGraphql extends Omit<IRole, 'users'> {
    users: IUser[]
}
"""

**Comment by nguyentb** [comment]
> > hi Chon, can you revise the pull request as I modify the tests accordingly
> 
> Hi, the tests are still not passing. Could you look into that?

Hi please get the latest version. the all tests get now passed

File: packages/itmat-interface/src/utils/mfa.js (hunk 3 above)
Comment (nguyentb) [reply packages/itmat-interface/src/utils/mfa.js:63]: > could you convert this to typescript

Sure I will do :D

**Comment by sou-chon** [comment]
> > > hi Chon, can you revise the pull request as I modify the tests accordingly
> > 
> > 
> > Hi, the tests are still not passing. Could you look into that?
> 
> Hi please get the latest version. the all tests get now passed

Thanks. but this PR has merge conflict. could you resolve them first

File: packages/itmat-job-executor/src/setupProxy.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+const { createProxyMiddleware } = require('http-proxy-middleware');
+
+module.exports = function (app) {
+    app.use(createProxyMiddleware('/graphql', {
+        ws: true,
+        target: 'http://localhost:3003',
+        changeOrigin: true
+    }));
+};
```
Comment (sou-chon) [inline packages/itmat-job-executor/src/setupProxy.js:9]: why does job-executor even have graphql endpoint?

Comment (sou-chon) [reply packages/itmat-job-executor/src/setupProxy.js:9]: also what is this commit doing in a PR for mfa?

Comment (nguyentb) [reply packages/itmat-job-executor/src/setupProxy.js:9]: I haven't wrote this code. As @fguitton created the branch for me, I think he will know about this :o

Comment (fguitton) [reply packages/itmat-job-executor/src/setupProxy.js:9]: As mentioned in my first comment (https://github.com/dsi-icl/itmat-broker/pull/484#issuecomment-622957655), you should rebase your changes ontop of the develop branch which @sou-chon has now fixed regarding the `react-script`/`lerna` incompatibility issue.

Comment (nguyentb) [reply packages/itmat-job-executor/src/setupProxy.js:9]: @fguitton so should I check out the develop branch and add the mfa function from the beginning? or any way better to modify the code in accordance with the develop branch? Thanks

**Comment by nguyentb** [comment]
> @nguyentb This branch is out of sync with the base develop. Also as I know that the react-script issue has been fixed in branch develop as @sou-chon push some changes, you might want to rebase you single commit directly off of there. This will avoid PR pollution with my commits which are not strictly related to the MFA 😉

I missed this comment :((

**Comment by fguitton** [comment]
> > @nguyentb This branch is out of sync with the base develop. Also as I know that the react-script issue has been fixed in branch develop as @sou-chon push some changes, you might want to rebase you single commit directly off of there. This will avoid PR pollution with my commits which are not strictly related to the MFA 😉
> I missed this comment :((

Not an issue. You can use a combination of `git checkout` and `git rebase` and some `--cherry-pick` to change the origin of commit e48f751 😉 .

**Comment by nguyentb** [comment]
@sou-chon @fguitton  I've already rebased the branch to merge the latest update from develop branch. All mfa-related features are now done. All test-cases are passed (132/132) Could you revise the pull request again?
Thanks :D

File: packages/itmat-interface/test/serverTests/_loginHelper.ts [hunk 5]
```diff
@@ -13,12 +15,14 @@ export function connectUser(agent) {
     return connectAgent(agent, 'standardUser', 'admin');
 }
 
-export function connectAgent(agent, user, pw) {
+export function connectAgent(agent, user, pw, otpSecret = "H6BNKKO27DPLCATGEJAZNWQV4LWOTMRA") {    
+    const otp = mfa.generateTOTP(otpSecret);
```
Comment (fguitton) [inline packages/itmat-interface/test/serverTests/_loginHelper.ts:19]: This probably doesn't belong here, My understanding is that it will generate a code that we expect to be good anyway. Furthermore, I do not see a test that verifies that the validation fails if no token is sent or an invalid one, this should be added to the test suite.

File: packages/itmat-interface/test/serverTests/users.test.ts [hunk 6]
```diff
@@ -420,24 +436,29 @@ describe('USERS API', () => {
             });
 
             /* getting the id of the created user from mongo */
-            const createdUserId = (await mongoClient
+            // const createdUserId = (await mongoClient
+            //     .collection(config.database.collections.users_collection)
+            //     .findOne({ username: 'testuser1' }, { projection: { id: 1 } })).id;
+
```
Comment (fguitton) [inline packages/itmat-interface/test/serverTests/users.test.ts:442]: Should the comment block still be present ?

File: packages/itmat-interface/test/serverTests/users.test.ts [hunk 7]
```diff
@@ -420,24 +436,29 @@ describe('USERS API', () => {
...
+            const createdUser = (await mongoClient
                 .collection(config.database.collections.users_collection)
-                .findOne({ username: 'testuser1' }, { projection: { id: 1 } })).id;
+                .findOne({ username: 'testuser1' }, { projection: { } }));
```
Comment (fguitton) [inline packages/itmat-interface/test/serverTests/users.test.ts:445]: Does it make sense to keep an empty projection ?

File: packages/itmat-interface/test/serverTests/users.test.ts [hunk 8]
```diff
@@ -604,7 +628,7 @@ describe('USERS API', () => {
                     query: print(EDIT_USER),
                     variables: {
                         id: 'fakeid2',
-                        password: 'admin',
+                        password: 'admin',
```
Comment (fguitton) [inline packages/itmat-interface/test/serverTests/users.test.ts:631]: There is a whole bunch of empty spaces added here

File: packages/itmat-ui-react/src/components/users/userDetails.tsx [hunk 9]
```diff
@@ -76,6 +76,7 @@ export const EditUserForm: React.FunctionComponent<{ user: (IUserWithoutToken &
                         </select></label><br /><br />
                     <label>Real name: <input type='text' value={inputs.realName} onChange={e => { setInputs({ ...inputs, realName: e.target.value }) }} /> </label><br /><br />
                     <label>Password:  <input type='password' value={inputs.password} onChange={e => { setInputs({ ...inputs, password: e.target.value }) }} /></label> <br /><br />
+                    <label>OTP Secret (readonly): <input type='text' readOnly value={inputs.otpSecret} /> </label><br /><br />
```
Comment (fguitton) [inline packages/itmat-ui-react/src/components/users/userDetails.tsx:79]: I don't think we should bring the OTP secret back to the UI anytime after generation. There is no immediate use for it I shouldn't think. What the intended purpose here ?

File: packages/itmat-interface/test/serverTests/_loginHelper.ts (hunk 5 above)
Comment (nguyentb) [reply packages/itmat-interface/test/serverTests/_loginHelper.ts]: Thank Florian for pointing it out. I've modified the code to clarify it better. The otpSecret for each users is randomly generated (except hardcoded for admin and standardUser at bootstrap).
I also add a new test case for the incorrect token :)

File: packages/itmat-interface/test/serverTests/users.test.ts (hunk 6 above)
Comment (nguyentb) [reply packages/itmat-interface/test/serverTests/users.test.ts]: done

File: packages/itmat-interface/test/serverTests/users.test.ts (hunk 8 above)
Comment (nguyentb) [reply packages/itmat-interface/test/serverTests/users.test.ts]: done

File: packages/itmat-ui-react/src/components/users/userDetails.tsx (hunk 9 above)
Comment (nguyentb) [reply packages/itmat-ui-react/src/components/users/userDetails.tsx:79]: at the moment I just wanna show it for putting into Google Authenticator app :D. Yeah, for later we can discuss how to send this secret to users :)

**Comment by sou-chon** [comment]
Hi, i know you changed the tests to fit the new functionality. but are there tests to test the OTP functions? like if the scenario where the OTP secrets are wrong
